### ISSUE TITLE
ngMessages is not working as expected

### ISSUE THREAD

**Issue by prakashjp123** [angular/material#6440]
I am using ngMessage with angular material. But it doesn't working as expected. I have my code [here](http://codepen.io/anon/pen/YwWLGW).

The problem is once the error message is displayed, when another new error is appeared and the existing error is disappeared then it breaks. For above example, click on the input and click outside of it. You will receive the required error. Now when you try inserting any character, the required error disappeared and the email error should appear. But this is not working.

**Comment by hewaroth** [comment]
```
<div ng-app="app" ng-controller="FormController" layout-padding>
  <form name="inputForm">
    <md-input-container>
      <label>Input Value</label>
      <input type="email" name="email" ng-required="true" ng-pattern="/^.+@.+\..+$/" ng-model="model.value" />
      <div ng-messages="inputForm.email.$error">
        <div ng-message-exp="['required', 'pattern']">E-mail is not valid</div>
      </div>
    </md-input-container> 
    <div>inputForm.inputValue.$error = {{inputForm.email.$error | json}}</div>
  </form>
</div>
```

**Comment by ebastuart** [comment]
I have the same issue with the newest update of material and angular

**Comment by gregorst** [comment]
Another issue here is that error message does not show initially for required input field.

**Comment by FrancescoRizzi** [comment]
on the last bit, I think it is by design... I have discussed this with users a number of times, and they don't seem to like being presented with error messages until they at least enter something, so the idea that ng-messages are hidden until the related field has been $touched seems good to me.

Messages that should be displayed even before the user interacts with a field seem to me to be labels or input hints.

**Comment by rsteenbergen** [comment]
Same issue here on ng-material 1.0.0, 1.0.1 and angular 1.4.6 and 1.4.8.

**Comment by gregorst** [comment]
@FrancescoRizzi: I thought so. At least we need a way to enable this.
And suggestion from @rsteenbergen is exactly what we need.  This is also described in docs.

**Comment by prakashjp123** [comment]
I think the hide animation works after the element appears into the display section. So it hides all messages. Is there any way to solve the issue temporarily?

**Comment by simonh1000** [comment]
@rsteenbergen thanks - that helped me

**Comment by ketanbhatt** [comment]
NgMessage behave erratically with angular-material. Here is a plunker: http://embed.plnkr.co/SWJFsOS3BuxYxePBQZ0C/

It works fine with just the `required` validation check, but if I add another, like `min` for instance, then sometimes the error message isn't shown at all.
Without `angular-material` it works fine.

**Form Without Angular-Material**
*Initial*
<img width="443" alt="screen shot 2016-01-22 at 4 37 53 pm" src="https://cloud.githubusercontent.com/assets/9729957/12509241/05c9be7a-c127-11e5-891d-5c866d7eca9a.png">

*When I enter a value < 400:*
<img width="401" alt="2" src="https://cloud.githubusercontent.com/assets/9729957/12509299/76eea69c-c127-11e5-9f7c-d85371c912c6.png">

**Form With Angular-Material**
*I enter a small value:*
<img width="358" alt="screen shot 2016-01-22 at 4 38 38 pm" src="https://cloud.githubusercontent.com/assets/9729957/12509316/9ab04c5c-c127-11e5-99ea-a37e9b1717a3.png">

*I erase everything:*
<img width="401" alt="screen shot 2016-01-22 at 4 38 31 pm" src="https://cloud.githubusercontent.com/assets/9729957/12509327/b32efecc-c127-11e5-998d-7378a87bc6c9.png">

*I enter a value greater than 400 and then erase everything:*
<img width="323" alt="screen shot 2016-01-22 at 4 38 51 pm" src="https://cloud.githubusercontent.com/assets/9729957/12509411/435239c4-c128-11e5-9e98-65f6d189c2d4.png">

**Comment by EladBezalel** [comment]
So this is resolved?

**Comment by joerideg** [comment]
Well @rsteenbergen provided a workaround, but is that the solution we want? Shouldnt it 'just work' with ngMessages instead of users having to add rsteenbergen's solution.

**Comment by wcjr** [comment]
I'm seeing the same problem and agree with @joerideg .  It is erratic so I hesitate to use the `md-hide` workaround to get it to work more often.

**Comment by langdonx** [comment]
Using `md-auto-hide="false"` isn't a great fix, since all errors will display by default, which can be overwhelming for someone filling out a new empty form.

**Comment by jimmyn** [comment]
I have the same problem here with angular 1.5.0 and angular-material 1.0.5. Also after I updated angular-material from 1.0.1 now the error pushes content down instead of appearing in md-errors-spacer

you can see it's even in official demo https://material.angularjs.org/latest/demo/input

**Comment by andrey-bahrachev** [comment]
Confirming this too (angular 1.5.0 + angular-material 1.0.5).

**Comment by jimmyn** [comment]
Any fix on this?

**Comment by pablozandona** [comment]
+1

**Comment by DanPii** [comment]
+1

**Comment by 2nise** [comment]
same here

**Comment by danielebuso** [comment]
+1

**Comment by pablozandona** [comment]
I resolved this adding 'ngMessages' in my modules ¬¬

**Comment by mbellezi** [comment]
+1

**Comment by adam-s** [comment]
+1 Not sure what the solution is here.

**Comment by manish2535** [comment]
+1

**Comment by julienmartin** [comment]
:+1:

**Comment by xiaoxiyao** [comment]
+1

**Comment by unfortunatetoast** [comment]
+1

**Comment by BeSaRa** [comment]
i'm trying workaround , i found the solution on CSS 
write this code after angular-material.css 

`.md-input.ng-invalid.ng-touched + .md-errors-spacer + [ng-messages] > div {
  opacity: 1 ;
  margin-top: 0;
}
.md-input.ng-invalid.ng-touched + .md-errors-spacer + [ng-messages] > div.ng-leave ,  .md-input.ng-invalid.ng-touched + .md-errors-spacer + [ng-messages] > div.ng-enter {
  position: absolute !important;
}

.md-input.ng-invalid.ng-touched + .md-errors-spacer + [ng-messages] > div.ng-leave  , .md-input.ng-invalid.ng-touched + .md-errors-spacer + [ng-messages] > div.ng-enter-active{
  opacity: 1 !important;
  margin-top: 0 !important;
}
.md-input.ng-invalid.ng-touched + .md-errors-spacer + [ng-messages] > div.ng-leave-active , , .md-input.ng-invalid.ng-touched + .md-errors-spacer + [ng-messages] > div.ng-enter  {
  opacity: 0 !important;
  margin-top: -100px !important;
}
`

i think this solution

**Comment by iamisti** [comment]
This issue has opened on Dec 22, do we expect the fix in this year or shall we wait another year for the fix?

**Comment by BeSaRa** [comment]
you can try this solution @iamisti
```
.md-input.ng-invalid.ng-touched + .md-errors-spacer + [ng-messages] > div {
opacity: 1 ;
margin-top: 0;
}
.md-input.ng-invalid.ng-touched + .md-errors-spacer + [ng-messages] > div.ng-leave , .md-input.ng-invalid.ng-touched + .md-errors-spacer + [ng-messages] > div.ng-enter {
position: absolute !important;
}

.md-input.ng-invalid.ng-touched + .md-errors-spacer + [ng-messages] > div.ng-leave , .md-input.ng-invalid.ng-touched + .md-errors-spacer + [ng-messages] > div.ng-enter-active{
opacity: 1 !important;
margin-top: 0 !important;
}
.md-input.ng-invalid.ng-touched + .md-errors-spacer + [ng-messages] > div.ng-leave-active , .md-input.ng-invalid.ng-touched + .md-errors-spacer + [ng-messages] > div.ng-enter {
opacity: 0 !important;
margin-top: -19px !important;
}
```

**Comment by tucq88** [comment]
```
Same issue here on ng-material 1.0.0, 1.0.1 and angular 1.4.6 and 1.4.8.

Update: adding md-auto-hide="false" solves my problems. I can now add ng-show="form2.username.$touched" to hide on page load and voila, I'm golden.
```
Thanks @rsteenbergen for this tricky solution. 
But with this solution, the message blinks at first time it displays (v1.0.7)

**Comment by ThomasBurleson** [comment]
This issue is closed as part of our ‘Surge Focus on Material 2' efforts.
        For details, see our [forum posting](http://bit.ly/1UhZyWs).

**Comment by ThomasBurleson** [comment]
This issue is closed as part of our ‘Surge Focus on Material 2' efforts.
      For details, see our forum posting @ http://bit.ly/1UhZyWs.

**Comment by ThomasBurleson** [comment]
This issue is closed as part of our ‘Surge Focus on Material 2' efforts.
      For details, see our forum posting @ http://bit.ly/1UhZyWs.

**Comment by nknowles88** [comment]
So first, I realize that this issue has be deprecated/closed as part of the _Surge Focus on Material 2_.
However, for those of you who want a proper fix read on.

But first, a little history for the _TL;DR_ folks.

This issue only happens when there are 2 or more `ng-message` containers within the `ng-messages` container.

@rsteenbergen provided a workaround (aka hack) for the div containing the `ng-messages` attribute directive that alleviated this issue:

> adding md-auto-hide="false" solves my problems. I can now add ng-show="form2.username.$touched" to hide on page load and voila, I'm golden.

This works, but it presents what I would call a 'jitter' issue. Which is basically, for lack of better explanation, when the same validation message has the ngMaterial _showMessage_ animation applied twice in quick succession.

As @joerideg said

> Well @rsteenbergen provided a workaround, but is that the solution we want? Shouldnt it 'just work' with ngMessages instead of users having to add rsteenbergen's solution.

Well here is the fix I have used for angular-1.5.8 using angular-material-1.1.0-rc.5.

In the `ngMaterial::input::ngMessageAnimation.enter` function, add the piece in **bold**

// If we have the md-auto-hide class, the md-input-invalid animation will fire, so we can skip
if (messages.hasClass('md-auto-hide') **&& messages.childElementCount === 1**) {
`  done();`
`  return;`
}

**TL;DR**
Apply the above change to the function _ngMessageAnimation()_ in
`node_modules/angular-material/angular-material.js`

**Comment by EladBezalel** [comment]
@nknowles88 thanks for the effort for those who still need this fix, **but**, i highly recommenced not to change the `node_modules` for two reasons:
1. When working with a team, that means every single person should do it
2. When we will release versions, your changes will be deleted

please, if you do want to help and get this issue resolve, suggest a PR and we will work to get it in.

**Comment by nknowles88** [comment]
Alright here is another solution that doesn't involve modifying vendor code.

- Create a css class that partially mimics what `md-auto-hide` is doing and place it in your application's css file. I called mine `hide-ng-message`
```css
md-input-container .hide-ng-message .md-input-message-animation:not(.ng-animate) {
  margin-top:-100px;
  opacity:0;
}
```
- On the container that has `ng-messages="myForm.myField.$error"`
  - add `hide-ng-message` class
  - add `md-auto-hide="false"`

```html
<md-input-container class="md-block">
  <label>Email</label>
  <input type="email" name="myField"
         ng-model="$ctrl.formData.myField"
         required/>
   <div class="hide-ng-message" role="alert"
        ng-messages="myForm.myField.$error" 
        md-auto-hide="false">
     <div ng-message="required">This is required!</div>
     <div ng-message="email">Must be a valid email address</div>
   </div>
</md-input-container>
```
This has the side effect of showing error messages as soon as the user starts typing in the input field. To retain the full effect of `md-auto-hide`, replace `md-auto-hide="false"` with `ng-class="{'md-auto-hide':myForm.myEmail.$untouched}"`.

---
@EladBezalel To make sure my prior suggestion didn't have any lingering effects, I deleted my entire node_modules folder and ran npm install to pull everything down again.

**Comment by EladBezalel** [comment]
@nknowles88 thanks for helping everyone encountering this issue,
IMO it would be even more helpful if you can actually solve this issue and suggest a PR :smile: 
thanks either way!

**Comment by nknowles88** [comment]
@EladBezalel This issue has been fixed already in the master branch 6a7a188 (which is 220 commits ahead of 1.1.0-rc.5).

I cloned the master, built the project and used the build artifacts to update/replace the `angular-material 1.1.0-rc.5 node_module` files that my project's build process uses.

I'm not familiar with this project's resolution process, but IMO this issue is a candidate for closure once a new release is made.
